components/eip681/src/parse.rs

```diff
@@ -0,0 +1,1449 @@
+
+#[derive(Debug, Snafu)]
+pub enum ParseError<'a> {
+    #[snafu(display("{}", nom::error::Error::to_string(&nom::error::Error {
+        input: *input, code: *code
+    })))]
+    Nom {
+        input: &'a str,
+        code: nom::error::ErrorKind,
+        other: Option<String>,
+    },
+
+    #[snafu(display(
+        "Expected at least {min} digits, saw {} before {}",
+        digits.len(),
+        input.split_at(10).0
+    ))]
+    DigitsMinimum {
+        min: usize,
+        digits: Vec<u8>,
+        input: &'a str,
+    },
+
+    #[snafu(display("Missing ENS name"))]
+    EnsMissing,
+
+    #[snafu(display("Not a domain"))]
+    EnsDomain,
+
+    #[snafu(display("{source}"))]
+    EnsNormalization {
+        source: ens_normalize_rs::ProcessError,
+    },
+
+    #[snafu(display(
+        "Invalid bit size. Expected {}..={} but saw {seen}",
+        range.start(),
+        range.end()
+    ))]
+    InvalidBitRange {
+        range: std::ops::RangeInclusive<u32>,
+        seen: u64,
+    },
+
+    #[snafu(display("Bit size must be a multiple of 8, saw {seen}"))]
+    InvalidBitSize { seen: u64 },
+
+    #[snafu(display("Invalid parameter value. Expected a number."))]
+    InvalidParameterValue,
+}
+
+impl<'a> nom::error::ParseError<&'a str> for ParseError<'a> {
+    fn from_error_kind(input: &'a str, code: nom::error::ErrorKind) -> Self {
+        NomSnafu {
+            input,
+            code,
+            other: None,
+        }
+        .build()
+    }
+
+    fn append(input: &'a str, code: nom::error::ErrorKind, other: Self) -> Self {
+        NomSnafu {
+            input,
+            code,
+            other: Some(other.to_string()),
+        }
+        .build()
+    }
+}
+
+impl<'a> From<ParseError<'a>> for nom::Err<ParseError<'a>> {
+    fn from(value: ParseError<'a>) -> Self {
+        if value.is_recoverable() {
+            nom::Err::Error(value)
+        } else {
+            nom::Err::Failure(value)
+        }
+    }
+}
+
+impl ParseError<'_> {
+    pub fn is_recoverable(&self) -> bool {
+        true
+    }
+}
+
+#[derive(Debug, Snafu, PartialEq)]
+pub enum ValidationError {
+    #[snafu(display("Exponent is too small, expected at least {expected}, saw {seen}"))]
+    SmallExponent { expected: usize, seen: u64 },
+
+    #[snafu(display("Number {seen} is too big and has saturated"))]
+    Saturation { seen: Number },
+
+    #[snafu(display("Could not decode url-encoded string: {source}"))]
+    UrlEncoding { source: std::str::Utf8Error },
+}
+
+/// Zero or more consecutive digits.
+///
+/// ```abnf
+/// *DIGIT
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Digits {
+    places: Vec<u8>,
+}
+
+impl core::fmt::Display for Digits {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for n in self.places.iter() {
+            f.write_fmt(format_args!("{n}"))?;
+        }
+        Ok(())
+    }
+}
+
+impl Digits {
+    pub const MAX_PLACES: usize = u64::MAX.ilog10() as usize;
```

Comment:
I think there's an off-by-one here.
```suggestion
    pub const MAX_PLACES: usize = (u64::MAX.ilog10() as usize) + 1;
```